pkg/assembler/backends/testing/hashEqual.go

```diff
@@ -17,113 +17,205 @@ package testing
+			Origin:        "testing backend",
+			Collector:     "testing backend",
+		})
+}
 
-	for _, a := range c.hashEquals {
-		if reflect.DeepEqual(a.Artifacts, artifacts) && a.Justification == justification {
-			return a
-		}
+func (c *demoClient) hashEqualByID(id uint32) (*hashEqualStruct, error) {
+	o, ok := c.index[id]
+	if !ok {
+		return nil, errors.New("could not find hashEqual")
 	}
-
-	newHashEqual := &model.HashEqual{
-		Justification: justification,
-		Artifacts:     artifacts,
-		Origin:        origin,
-		Collector:     collector,
+	a, ok := o.(*hashEqualStruct)
+	if !ok {
+		return nil, errors.New("not a hashEqual")
 	}
-	c.hashEquals = append(c.hashEquals, newHashEqual)
-	return newHashEqual
+	return a, nil
 }
 
+// Ingest HashEqual
 func (c *demoClient) IngestHashEqual(ctx context.Context, artifact model.ArtifactInputSpec, equalArtifact model.ArtifactInputSpec, hashEqual model.HashEqualInputSpec) (*model.HashEqual, error) {
 
-	collectedArt, err := c.Artifacts(ctx, &model.ArtifactSpec{Algorithm: &artifact.Algorithm, Digest: &artifact.Digest})
+	aInt1, err := c.artifactByKey(artifact.Algorithm, artifact.Digest)
 	if err != nil {
-		return nil, err
-	}
-	if len(collectedArt) != 1 {
-		return nil, gqlerror.Errorf(
-			"IngestHashEqual :: multiple artifacts found")
+		return nil, gqlerror.Errorf("IngestHashEqual :: Artifact not found")
 	}
-
-	collectedEqualArt, err := c.Artifacts(ctx, &model.ArtifactSpec{Algorithm: &equalArtifact.Algorithm, Digest: &equalArtifact.Digest})
+	aInt2, err := c.artifactByKey(equalArtifact.Algorithm, equalArtifact.Digest)
 	if err != nil {
-		return nil, err
+		return nil, gqlerror.Errorf("IngestHashEqual :: Artifact not found")
 	}
-	if len(collectedEqualArt) != 1 {
-		return nil, gqlerror.Errorf(
-			"IngestHashEqual :: multiple artifacts found")
+	artIDs := []uint32{aInt1.id, aInt2.id}
+	sort.Slice(artIDs, func(i, j int) bool { return artIDs[i] < artIDs[j] })
+
+	// Search backedges for existing.
+	searchHEs := slices.Clone(aInt1.hashEquals)
+	searchHEs = append(searchHEs, aInt2.hashEquals...)
+
+	for _, he := range searchHEs {
+		h, err := c.hashEqualByID(he)
+		if err != nil {
+			return nil, gqlerror.Errorf(
+				"IngestHashEqual :: Bad hashEqual id stored on existing artifact: %s", err)
+		}
+		if h.justification == hashEqual.Justification &&
+			h.origin == hashEqual.Origin &&
+			h.collector == hashEqual.Collector &&
+			slices.Equal(h.artifacts, artIDs) {
+			return c.convHashEqual(h), nil
+		}
 	}
 
-	return c.registerHashEqual(
-		[]*model.Artifact{collectedArt[0], collectedEqualArt[0]},
-		hashEqual.Justification,
-		hashEqual.Origin,
-		hashEqual.Collector), nil
+	he := &hashEqualStruct{
+		id:            c.getNextID(),
+		artifacts:     artIDs,
+		justification: hashEqual.Justification,
+		origin:        hashEqual.Origin,
+		collector:     hashEqual.Collector,
+	}
+	c.index[he.id] = he
+	c.hashEquals = append(c.hashEquals, he)
+	aInt1.hashEquals = append(aInt1.hashEquals, he.id)
```

Comment:
can do a similar helper function like with package and source to append and get.